packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -48,43 +50,85 @@ export class AssetStaging extends Construct {
   public readonly sourcePath: string;
 
   /**
-   * A cryptographic hash of the source document(s).
+   * A cryptographic hash of the asset.
+   *
+   * @deprecated see `assetHash`.
    */
   public readonly sourceHash: string;
 
+  /**
+   * A cryptographic hash of the asset.
+   */
+  public readonly assetHash: string;
+
   private readonly fingerprintOptions: FingerprintOptions;
 
   private readonly relativePath?: string;
 
+  private readonly bundleDir?: string;
+
   constructor(scope: Construct, id: string, props: AssetStagingProps) {
     super(scope, id);
 
+    if (props.assetHash && props.assetHashType && props.assetHashType !== AssetHashType.CUSTOM) {
+      throw new Error(`Cannot specify \`${props.assetHashType}\` for \`assetHashType\` when \`assetHash\` is specified. Use \`CUSTOM\` or leave \`undefined\`.`);
+    }
+
     this.sourcePath = props.sourcePath;
     this.fingerprintOptions = props;
-    this.sourceHash = FileSystem.fingerprint(this.sourcePath, props);
+
+    if (props.bundling) {
+      this.bundleDir = this.bundle(props.bundling);
+    }
+
+    const hashType = props.assetHash ? AssetHashType.CUSTOM : props.assetHashType ?? AssetHashType.SOURCE;
+    this.assetHash = this.calculateHash(hashType, props.assetHash);
```

Comment:
Technically you are missing a validation here (`assetHash` is defined but `assetHashType` is not CUSTOM).

Reply:
It's on L73-L75 above, will move it closer

Reply:
I would just move the validation into `calculateHash` and pass in `props`.